library/pkparse.c

```diff
@@ -1253,14 +1253,59 @@ int mbedtls_pk_parse_public_key( mbedtls_pk_context *ctx,
 #if defined(MBEDTLS_PEM_PARSE_C)
     size_t len;
     mbedtls_pem_context pem;
-
+    const mbedtls_pk_info_t *pk_info;
     mbedtls_pem_init( &pem );
-
+#if defined(MBEDTLS_RSA_C)
     /* Avoid calling mbedtls_pem_read_buffer() on non-null-terminated string */
     if( keylen == 0 || key[keylen - 1] != '\0' )
         ret = MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT;
     else
         ret = mbedtls_pem_read_buffer( &pem,
+                               "-----BEGIN RSA PUBLIC KEY-----",
+                               "-----END RSA PUBLIC KEY-----",
+                               key, NULL, 0, &len );
+
+    if( ret == 0 )
+    {
+        p = pem.buf;
+        if( ( pk_info = mbedtls_pk_info_from_type( MBEDTLS_PK_RSA ) ) == NULL )
+            return( MBEDTLS_ERR_PK_UNKNOWN_PK_ALG );
+
+        if( ( ret = mbedtls_pk_setup( ctx, pk_info ) ) != 0 )
+            return( ret );
+
+        if ( ( ret = pk_get_rsapubkey( &p, p + pem.buflen, mbedtls_pk_rsa( *ctx ) ) ) != 0 )
+            mbedtls_pk_free( ctx );
+        mbedtls_pem_free( &pem );
+        return( ret );
+    }
+    else if( ret != MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT )
+    {
+        mbedtls_pem_free( &pem );
+        return( ret );
+    }
+
+    if( ( pk_info = mbedtls_pk_info_from_type( MBEDTLS_PK_RSA ) ) == NULL )
+          return( MBEDTLS_ERR_PK_UNKNOWN_PK_ALG );
+
+    if( ( ret = mbedtls_pk_setup( ctx, pk_info ) ) != 0 )
+        return( ret );
+
+    p =  (unsigned char *) key;
+    ret = pk_get_rsapubkey( &p, p + keylen, mbedtls_pk_rsa( *ctx ) );
+    if ( ret == 0 )
```

Comment:
This way, a PEM encoded generic public key (guarded by `BEGIN/END PUBLIC KEY`) will not be immediately recognized as such, but instead will be passed to the DER-parsing routine `pk_get_rsapubkey`. While the latter will gracefully fail and lead to the key being subsequently correctly categorized, I think we shouldn't call any DER parsing routines on PEM encoded keys, as the latter can be easily recognized as such.

Concretely: Can we align the code with the structure that the private key parsing function `mbedtls_pk_parse_key`, i.e. go through all possible PEM formats first and only afterwards try the DER parsing routines?

Reply:
I thought about doing that, but this adds more `#if defined(MBEDTLS_RSA_C)` which I am not too fond of adding too much.
In addition, even if we send the pkcs8 PEM formatted file to the pkcs1 der parsing, it will fail, and then we continue to do the PEM parsing.
In my opinion, it's more readable to first try pkcs1 parsing, and then do the pkcs8 parsing

Reply:
My main argument is that I'd like to have `mbedtls_pk_parse_key` and `mbedtls_pk_parse_public_key` follow the same strategy, regardless of which one that is (although I prefer to do the PEM vs. DER distinction first) Let's wait for another opinion maybe?

Reply:
I am starting to tend to agree with, to start with PEM, and then DER, since this is how `mbedtls_pk_parse_key` is implemented, but this means bigger change on the code. It means to modify(in pseudo code)
~~~
#if defined(MBEDTLS_PEM_PARSE_C)
if  parse PEM succeeds:
    do stuff
    parse DER
    return ret
#endif /* MBEDTLS_PEM_PARSE_C */
parse DER
~~~
Following this, I will also need to change the regular parsing, which might be a bit out of scope of this PR